src/functions.php

```diff
@@ -478,23 +478,48 @@ function is_settled(PromiseInterface $promise)
 function coroutine(callable $generatorFn)
 {
     $generator = $generatorFn();
-    return __next_coroutine($generator->current(), $generator)->then();
+    $state = new \stdClass;
```

Comment:
Rather than relying on an instance of `stdClass`, I would suggest creating creating a class to maintain coroutine state. You could even go so far as to encapsulate coroutine internals in a class that implements PromisorInterface.